docs/sdk/src/reference_docs/tokens_in_substrate.rs

```diff
@@ -0,0 +1,97 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Tokens in Substrate
+//!
+//! This reference doc serves as a high-level overview of the token-related logic in Substrate and
+//! how to properly apply it to your usecase.
+//!
+//! On completion of reading this doc, you should have a good understanding of
+//! - The distinction between token traits and trait implementations in Substrate, and why this
+//!   distinction is helpful
+//! - Token-related traits avaliable in Substrate
+//! - Token-related trait implementations in Substrate
+//! - How to choose the right trait or trait implementation for your use case
+//! - Where to go next
+//!
+//! ## Traits and Trait Implementations
+//!
+//! Broardly speaking, token logic in Substrate can be divided into two categories: traits and
+//! trait implementations.
+//!
+//! **Traits** define common interfaces that types of token should implement. For example, the
+//! [`frame_support::traits::fungible::Inspect`] trait specifies an interface for *inspecting*
+//! token state such as the total issuance of the token, the balance of individual accounts, etc.
+//!
+//! **Trait implementations** are concrete implementations of these traits. For example, one of the
+//! many traits [`pallet_balances`] implements is [`frame_support::traits::fungible::Inspect`].
+//!
+//! The distinction between traits and trait implementations is helpful because it allows pallets
+//! and other logic to be generic over their dependencies, avoiding cumbersome pallet tight
+//! coupling.
+//!
+//! To illustrate this with an example let's consider [`pallet_preimage`]. This pallet takes a
+//! deposit in exchange for storing some preimage for use later. A naive implementation of the
+//! pallet may use [`pallet_balances`] as a dependency, and directly call the methods exposed by
+//! [`pallet_balances`] to reserve and unreserve deposits. This approach works well, until someone
+//! has a usecase requiring that an asset from a different pallet such as [`pallet_assets`] is
+//! used for the deposit. Rather than tightly couple [`pallet_preimage`] to [`pallet_balances`] AND
+//! [`pallet_assets`], along with every other token type pallet a user could possibly specify,
+//! [`pallet_preimage`] does not specify a concrete pallet as a dependency but instead accepts any
+//! dependency which implements the [`frame_support::traits::tokens::currency::ReservableCurrency`]
```

Comment:
We should probably migrate `pallet_preimage` to use fungible so we don't have to use the old traits as the example

Reply:
Yeah.. There's an open PR for it, waiting for MBM to write a migration.